### PR TITLE
WP Scripts: Build block.json viewModule

### PR DESCRIPTION
**Pull request by sirreal**
## What?

Add handling of `viewModule` field in block.json when using `wp-scripts` `build` or `start`.

Part of https://github.com/WordPress/gutenberg/issues/57492

## Why?

As the Modules API matures, our tooling should support it.

## How?

We add an option to clearly mark this an an experimental feature: `wp-scripts build --experimental-modules`.

To support modules with webpack, it was necessary to run a multi-compilation. We use an array of webpack configurations instead of a single webpack configuration. This is because module compilation is an option at the compilation level and cannot be set for specific entrypoints.

If the `--experimental-modules` option is found, we use an environment variable to change the `webpack.config` export to return an array `[ scriptWebpackConfig, moduleWebpackConfig ]`. Without the experimental option, the `webpack.config` export is the same except for changes that should not impact the script build at all. Consumers should be able to continue extending this config without noticing any differences. I introduced a snapshot test to compare the results, the difference is currently at 20dff00fef6. I did subsequently remove this test as it doesn't seem to have long-term value.

Notable changes to the existing webpack config: **None.**

- ~`experiments.outputModule` webpack option is set to `true`. This _enables_ module compilation, but module compilation is controled by other options (`output.module`). This should not impact existing builds.~

As mentioned above, if we find the `WP_EXPERIMENTAL_MODULES` environment variable (set by adding the `--experimental-modules` option on the CLI) then the webpack config will export an array of configs.

If we run in experimental modules mode, the `CleanWebpackPlugin` could remove assets that should be included. This is because both compilations output to the same output directory. Ideally, each compilation would have control over its output directory and not share them. That's difficult and I don't have a good solution at this time given that we do want to share other assets like stylesheets. Additionally, if we were to output to different directories, the block.json `file:…` rewriting would need to account for that.

I don't have a convincing solution for either using multiple directories or preventing assets from being clobbered at this time. Given that this is clearly flagged as experimental, I believe that's acceptable and we can improve things if a convincing solution presents itself.

Since this is experimental, I wanted avoided adding more webpack configuration files. I don't want to expose more files that could folks require and extend. We could consider [adding an `exports` field to this package](https://nodejs.org/api/packages.html#exports) to control access to certain things, but that's likely a breaking change I didn't want to undertake here.

## Testing Instructions

To test this, you'll want to check this out and link a project that depends on `@wordpress/scripts` to use this PR.

If you use `wp-scripts build` or `wp-scripts start`, there should be no significant changes with the current behavior.

You can test the experimental build by adding a `viewModule` field to block.json in a project, then running a script like `wp-scripts -- build --experimental-modules` to build it.

I've been running the experimental build of https://github.com/sirreal/wp-scripts-interactivity-module-demo (the packages is linked to my development copy locally). The build produces the expected results with both scripts and modules. With #57437 (handling `viewModule` in block registration), this results in a working plugin. Here's the build ouptut:

<details>

<summary><code>block.json</code></summary>

```json
{
  "$schema": "https://schemas.wp.org/trunk/block.json",
  "apiVersion": 3,
  "title": "Jon's cool counter",
  "name": "jon/the-block",
  "editorScript": "file:./index.js",
  "viewModule": "file:./view.js",
  "render": "file:./render.php",
  "category": "text"
}
```

</details>

<details>

<summary><code>index.asset.php</code></summary>

```php
<?php return array('dependencies' => array('react', 'wp-blocks'), 'version' => '04a7cb7df938528c50ee');
```

</details>

<details>

<summary><code>index.js</code></summary>

```js
const e=window.React,t=window.wp.blocks,o=JSON.parse('{"$schema":"https://schemas.wp.org/trunk/block.json","apiVersion":3,"title":"Jon\'s cool counter","name":"jon/the-block","editorScript":"file:./index.tsx","viewModule":"file:./view.mts","render":"file:./render.php","category":"text"}');t.registerBlockType(o,{edit:()=>(0,e.createElement)("div",null,"counter"),save:()=>(0,e.createElement)("div",null,"counter")});
```

</details>

<details>

<summary><code>render.php</code></summary>

```php
<div data-wp-interactive='{"namespace":"jon/the-block"}' data-wp-context='{"val":0}'>
	<input readonly type='number' data-wp-bind--value='context.val' value='0' />
	<button data-wp-on--click='actions.inc'>+</button>
	<button data-wp-on--click='actions.dec'>-</button>
</div>
```

</details>

<details>

<summary><code>view.asset.php</code></summary>

```php
<?php return array('dependencies' => array(array('id' => '@wordpress/interactivity', 'type' => 'dynamic')), 'version' => 'c9463319dadd37e31637', 'type' => 'module');
```

</details>

<details>

<summary><code>view.js</code></summary>

```js
var e={990:e=>{e.exports=import("@wordpress/interactivity")}},t={};function n(o){var r=t[o];if(void 0!==r)return r.exports;var i=t[o]={exports:{}};return e[o](i,i.exports,n),i.exports}(()=>{async function e(){if(e.init)return;e.init=!0;const t=await Promise.resolve().then(n.bind(n,990));t.store("jon/the-block",{actions:{inc(){t.getContext().val+=1},dec(){t.getContext().val-=1}}})}document.querySelector('[data-wp-interactive*=\'"namespace":"jon/the-block"\']').addEventListener("mouseenter",e,{once:!0}),document.querySelector('[data-wp-interactive*=\'"namespace":"jon/the-block"\']').addEventListener("focusin",e,{once:!0})})();
```

</details>

### CODE REVIEW

**Comment by sirreal** [comment]
I spent some time thinking about the problem of colliding assets and I think we should explore changing the output directories of script and module assets. This could mean things like CSS duplication, but I think we're unlikely to have both `script` and `viewModule` assets that load on a site's frontend and expect to share CSS assets. If they do, some duplicated CSS assets seem preferable to potentially broken assets.

That should make the cleaning strategies simpler as builds would have more control over their output directories.

I think something like this would be good, but I do want to do some testing:

```
build/
  script/editor.js
  script/…other-assets-from-script-build
  module/view.js
  render.php
  block.json
  …other-copied-assets
```

That doesn't need to block review of this PR, it can be landed and we can iterate on the output strategy.

File: packages/scripts/utils/config.js [hunk 1]
```diff
@@ -108,7 +112,9 @@ const hasPostCSSConfig = () =>
  */
 const getWebpackArgs = () => {
 	// Gets all args from CLI without those prefixed with `--webpack`.
-	let webpackArgs = getArgsFromCLI( [ '--webpack' ] );
+	let webpackArgs = getArgsFromCLI( [ '--webpack' ] ).filter(
```
Comment (gziolo) [inline packages/scripts/utils/config.js]: Would the following work, too?
```suggestion
	let webpackArgs = getArgsFromCLI( [ '--webpack', '--experimental-modules' ] );
```

File: packages/scripts/config/webpack.config.js [hunk 2]
```diff
@@ -245,21 +254,69 @@ const config = {
...
+// WP_DEVTOOL global variable controls how source maps are generated.
+// See: https://webpack.js.org/configuration/devtool/#devtool.
+if ( process.env.WP_DEVTOOL ) {
+	baseConfig.devtool = process.env.WP_DEVTOOL;
+}
+
+if ( ! isProduction ) {
+	// Set default sourcemap mode if it wasn't set by WP_DEVTOOL.
+	baseConfig.devtool = baseConfig.devtool || 'source-map';
+	baseConfig.devServer = {
```
Comment (gziolo) [inline packages/scripts/config/webpack.config.js]: Will two configs integrate seamlessly with a single dev server that uses shared setup?

File: packages/scripts/config/webpack.config.js [hunk 3]
```diff
@@ -324,45 +387,46 @@ const config = {
...
+				type: 'module',
 			},
 		},
+
+		plugins: [
+			new webpack.DefinePlugin( {
+				// Inject the `SCRIPT_DEBUG` global, used for development features flagging.
+				SCRIPT_DEBUG: ! isProduction,
+			} ),
+			new RenderPathsPlugin(),
```
Comment (gziolo) [inline packages/scripts/config/webpack.config.js]: This plugin ensures PHP files referenced in block.json get copied. It ‘s probably fine to have it only in the scripts config.

File: packages/scripts/config/webpack.config.js [hunk 4]
```diff
@@ -324,45 +387,46 @@ const config = {
...
+			// The WP_BUNDLE_ANALYZER global variable enables a utility that represents
+			// bundle content as a convenient interactive zoomable treemap.
+			process.env.WP_BUNDLE_ANALYZER && new BundleAnalyzerPlugin(),
+			// MiniCSSExtractPlugin to extract the CSS thats gets imported into JavaScript.
+			new MiniCSSExtractPlugin( { filename: '[name].css' } ),
+			// React Fast Refresh.
+			hasReactFastRefresh && new ReactRefreshWebpackPlugin(),
```
Comment (gziolo) [inline packages/scripts/config/webpack.config.js:423]: I don’t think React Fast Refresh will work out of the box for viewModule.

File: packages/scripts/config/webpack.config.js (hunk 2 above)
Comment (sirreal) [reply packages/scripts/config/webpack.config.js]: Good spot, this needs to be configured with different ports.

File: packages/scripts/utils/config.js (hunk 1 above)
Comment (sirreal) [reply packages/scripts/utils/config.js]: That works. I avoided it because `getArgsFromCLI` takes an array of "excluded prefixes" and that's not really a prefix. Under the hood it's excluding strings by `startWith` so it should be fine.

File: packages/scripts/config/webpack.config.js (hunk 2 above)
Comment (sirreal) [reply packages/scripts/config/webpack.config.js]: It looks like devServer doesn't really work with multiple compilations: https://github.com/webpack/webpack-cli/issues/2408

I'll only add this to 1 config.

File: packages/scripts/config/webpack.config.js (hunk 4 above)
Comment (sirreal) [reply packages/scripts/config/webpack.config.js:423]: I think it may work, but I'm not really sure how to test. Interactivity uses preact, which I don't think is compatible with the react refresh work. I saw this issue about refresh support for preact https://github.com/preactjs/preact/issues/2184

We probably just remove this for now for modules.

File: lib/experimental/modules/class-gutenberg-modules.php [hunk 5]
```diff
@@ -280,3 +280,191 @@ function gutenberg_dequeue_module( $module_identifier ) {
...
+/**
+ * Add module fields from block metadata to WP_Block_Type settings.
+ *
+ * This filter allows us to register modules from block metadata and attach additional fields to
+ * WP_Block_Type instances.
+ *
+ * @param array $settings Array of determined settings for registering a block type.
+ * @param array $metadata Metadata provided for registering a block type.
+ */
+function gutenberg_filter_block_type_metadata_settings_register_modules( $settings, $metadata = null ) {
```
Comment (luisherranz) [inline lib/experimental/modules/class-gutenberg-modules.php]: These are meant to be removed and merged on their own PR, right?

- https://github.com/WordPress/gutenberg/pull/57437

File: packages/scripts/config/webpack.config.js [hunk 6]
```diff
@@ -110,14 +116,17 @@ const cssLoaders = [
...
 	mode,
 	target,
-	entry: getWebpackEntryPoints,
 	output: {
 		filename: '[name].js',
 		path: resolve( process.cwd(), 'build' ),
 	},
+	experiments: {
+		outputModule: true,
+	},
```
Comment (luisherranz) [inline packages/scripts/config/webpack.config.js]: Why is this on the `baseConfig` instead of `moduleConfig`? Couldn't it affect the plugins extending the Webpack config like this, or is it innocuous?

```js
const defaultConfig = require("@wordpress/scripts/config/webpack.config");

module.exports = {
  // defaultConfig contains baseConfig, so `experiments.outputModule` is now here.
  ...defaultConfig,

  // some extra config...
};
```

**Review by luisherranz — APPROVED** [review]
LGTM! 👏 👏 

Tested in conjunction with https://github.com/WordPress/gutenberg/pull/57437 and everything worked as expected.

I only have a couple of comments, but I think this is ready to merge 🙂

File: lib/experimental/modules/class-gutenberg-modules.php (hunk 5 above)
Comment (sirreal) [reply lib/experimental/modules/class-gutenberg-modules.php]: Yes, my mistake. I've removed these changes from this PR.

File: packages/scripts/config/webpack.config.js (hunk 6 above)
Comment (sirreal) [reply packages/scripts/config/webpack.config.js]: As far as I can tell, by itself this doesn't change anything about how webpack works. It enables the module output experiment, but that's controlled by other options like `output.module`.

This seemed like a harmless, simple change, but there's no real justification to do it here. I'll move it to the module config.

**Comment by sirreal** [comment]
`viewModule` probably needs a dev note, but we'll cover that as part of #57437 or maybe a broader module announcement.

**Comment by sirreal** [comment]
> I spent some time thinking about the problem of colliding assets and I think we should explore changing the output directories of script and module assets. This could mean things like CSS duplication, but I think we're unlikely to have both `script` and `viewModule` assets that load on a site's frontend and expect to share CSS assets. If they do, some duplicated CSS assets seem preferable to potentially broken assets.
> 
> That should make the cleaning strategies simpler as builds would have more control over their output directories.
> 
> I think something like this would be good, but I do want to do some testing:
> 
> ```
> build/
>   script/editor.js
>   script/…other-assets-from-script-build
>   module/view.js
>   render.php
>   block.json
>   …other-copied-assets
> ```
> 
> That doesn't need to block review of this PR, it can be landed and we can iterate on the output strategy.

I put together a proposal in https://github.com/WordPress/gutenberg/pull/57761 that outputs different assets to different directories, but I'm not completely happy with it.

I now think that there is less likelihood of assets colliding because assets are named based on files in the src directory. Different assets can't have the same name because they would be the same src file (technically they probably could, but that's unlikely to happen in reality at this time).

Let's see what happens while folks use module builds and whether there's a problem in practice.

File: packages/scripts/config/webpack.config.js [hunk 7]
```diff
@@ -317,52 +380,55 @@ const config = {
 		process.env.WP_BUNDLE_ANALYZER && new BundleAnalyzerPlugin(),
 		// MiniCSSExtractPlugin to extract the CSS thats gets imported into JavaScript.
 		new MiniCSSExtractPlugin( { filename: '[name].css' } ),
-		// React Fast Refresh.
```
Comment (gziolo) [inline packages/scripts/config/webpack.config.js:320]: It looks like React Fast Refresh was misplaced. It should remain here for the scripts configuration as it works withe the existing Babel config and the dev server.

It won’t work with the modules config as it’s missing proper Babel handling, the dev server integration. In addition, I’m not sure whether it’s even possible to make it work with Preact.

File: packages/scripts/config/webpack.config.js (hunk 4 above)
Comment (gziolo) [reply packages/scripts/config/webpack.config.js:423]: See https://github.com/WordPress/gutenberg/pull/57461/files#r1449177817.

File: packages/scripts/config/webpack.config.js (hunk 7 above)
Comment (sirreal) [reply packages/scripts/config/webpack.config.js:320]: You're absolutely correct, thanks: https://github.com/WordPress/gutenberg/pull/57777

**Comment by Tropicalista** [comment]
How to change the previously working webpack additional config?

I mean I was using something like this to add additional entrypoint:

```
const defaultConfig = require( '@wordpress/scripts/config/webpack.config' );
const path = require( 'path' );

module.exports = {
	...defaultConfig,
	entry: {
		...defaultConfig.entry(),
		navigation: [
			path.resolve( __dirname, 'src', 'navigation', 'index.js' ),
		],
	},
	output: {
		filename: '[name].js',
		path: path.resolve( process.cwd(), 'build' ),
		clean: true,
	},
};
```

Now this is not working anymore. I got the error `TypeError: defaultConfig.entry is not a function`.

Can you please explain how to handle module compilation with multiple entrypoint?

File: packages/scripts/config/webpack.config.js [hunk 8]
```diff
@@ -245,21 +251,72 @@ const config = {
...
+		use: require.resolve( 'source-map-loader' ),
+		enforce: 'pre',
+	} );
+}
+
+/** @type {webpack.Configuration} */
+const scriptConfig = {
+	...baseConfig,
+
+	entry: getWebpackEntryPoints( 'script' ),
```
Comment (gziolo) [inline packages/scripts/config/webpack.config.js:284]: This needs to be refactored back to a callback function. There are two reasons:
- compatibility with the existing projects that override entry points
- more importantly, entry points are recomputed in the watch mode anymore

**Comment by sirreal** [comment]
> How to change the previously working webpack additional config?
> …
> Can you please explain how to handle module compilation with multiple entrypoint?

Thanks for the report @Tropicalista. These changes were not intended the modify the webpack config (unless you've opted in to experimental module compilation). I'll investigate.

File: packages/scripts/config/webpack.config.js (hunk 8 above)
Comment (sirreal) [reply packages/scripts/config/webpack.config.js:284]: It is a callback, the signature changed so this now returns the callback:

https://github.com/WordPress/gutenberg/blob/05b6bd76b6d8ca2a742601ebbd2f79bf582dd016/packages/scripts/utils/config.js#L188-L202

I'm investigating the problem now.

**Comment by sirreal** [comment]
> Can you please explain how to handle module compilation with multiple entrypoint?

I'm sorry, I missed that this was for _module_ compilation, so the build is being run with the `--experimental-modules` flag.

@Tropicalista With experimental module compilation we don't support using the webpack configuration directly for extension like in your example. Module compilation via this package is experimental and we need to be able to change or remove the module configuration. In the future, module compilation may not even be handled via webpack.

The main issue for discussing this module integration is https://github.com/WordPress/gutenberg/issues/57492. Not considering the module webpack config to be public is [one of the decisions we made at least for this first version](https://github.com/WordPress/gutenberg/issues/57492#issuecomment-1875619320):

> No webpack config for module compilation will be considered public at this time for consumers to extend.

File: packages/scripts/config/webpack.config.js (hunk 8 above)
Comment (sirreal) [reply packages/scripts/config/webpack.config.js:284]: > more importantly, entry points are recomputed in the watch mode anymore

Do you mean if I change block.json (to add another file) when `wp-scripts start --experimental-modules` is running, the build doesn't pick it up?

I don't think this has changed at all for script builds or for the script compilation of module builds. I believe the copy webpack plugin is adding the block.json files to the build which makes webpack observe those changes.

Because the modules compilation doesn't include the copy plugins for block.json, when block.json changes the module compilation won't pick it up. I'll look into how to make it observe those changes.

Comment (gziolo) [reply packages/scripts/config/webpack.config.js:284]: It looks like it should work correctly for scripts when the is no experimental flag provided 👍

Comment (sirreal) [reply packages/scripts/config/webpack.config.js:284]: Here's a PR to add `block.json` files as webpack file dependencies so they're watched from the module build: https://github.com/WordPress/gutenberg/pull/57927
